alexandria/skip_graph.py

```diff
@@ -0,0 +1,303 @@
+        elif left is None:
+            right.set_left_neighbor(level, None)
+        elif right is None:
+            left.set_right_neighbor(level, None)
+        else:
+            left.set_right_neighbor(level, right.key)
+            right.set_left_neighbor(level, left.key)
+
+    def _search_insert_point(self,
+                             key: Key,
+                             current: SGNode,
+                             level: int) -> Tuple[Optional[SGNode], Optional[SGNode]]:
+        if key == current.key:
+            raise AlreadyPresent
+        elif key > current.key:
+            for at_level, right_key in current.iter_down_right_levels(level):
+                if right_key is None:
+                    continue
+
+                if key >= right_key:
+                    right_neighbor = self.nodes[right_key]
+                    return self._search_insert_point(key, right_neighbor, at_level)
+            else:
+                right_neighbor_key = current.get_right_neighbor(0)
+                if right_neighbor_key is None:
+                    return current, None
+                right_neighbor = self.nodes[right_neighbor_key]
+                return current, right_neighbor
+        elif key < current.key:
+            for at_level, left_key in current.iter_down_left_levels(level):
+                if left_key is None:
+                    continue
+
+                if key <= left_key:
+                    left_neighbor = self.nodes[left_key]
+                    return self._search_insert_point(key, left_neighbor, at_level)
+            else:
+                left_neighbor_key = current.get_left_neighbor(0)
+                if left_neighbor_key is None:
+                    return None, current
+                left_neighbor = self.nodes[left_neighbor_key]
+                return left_neighbor, current
+        else:
+            raise Exception("Invariant")
+
+    def delete(self, key: Key, current: SGNode) -> None:
+        self.logger.debug('Deleting: %d', key)
+        node = self.search(key, current)
+
+        left: Optional[SGNode]
+        right: Optional[SGNode]
+
+        for level in range(node.max_level, -1, -1):
+            left_neighbor_key = node.get_left_neighbor(level)
+            if left_neighbor_key is None:
+                left = None
+            else:
+                left = self.nodes[left_neighbor_key]
+
+            right_neighbor_key = node.get_right_neighbor(level)
+            if right_neighbor_key is None:
+                right = None
+            else:
+                right = self.nodes[right_neighbor_key]
+
+            if left is not None or right is not None:
+                self._link_nodes(left, right, level)
+
+        self.nodes.pop(key)
+
+    def search(self, key: Key, current: SGNode) -> None:
+        self.logger.debug("Searching: %d", key)
+        return self._search(key, current, current.max_level)
+
+    def _search(self, key: Key, current: SGNode, level: int) -> SGNode:
```

Comment:
I'm not sure exactly how to do it, but you might be able to unify this with `_search_insert_point`, these two methods have a lot in common!

Maybe `_find_node` could return `left_neighbor, current, right_neighbor`. `_search_insert_point` interprets that result to either throw `AlreadyPresent` or return the neighbors. `_search` throws away the neighbors and throws `NotFound` or returns the node.

Reply:
I'm now exposing this as a top level API `GraphAPI.find`.  It'll be a necessary external API for doing range queries.